## 06-Temporal Validation and Comprative analysis

**Research Question 3**

**Which neighbourhood-level socioeconomic indicators changed significantly between 2011 and 2016 across Toronto, and what was the magnitude and direction of those changes?***

### Objective

This notebook extends the primary analysis by incorporating the 2011 Toronto Neighbourhood Profiles dataset to perform a temporal comparison with the 2016 machine learning-ready dataset.

The objectives are to evaluate the compatibility of the two datasets, examine changes in key socioeconomic indicators between 2011 and 2016, and assess whether the socioeconomic relationships identified in the primary analysis remain consistent over time.

Unlike the previous notebooks, which focused on modelling neighbourhood socioeconomic vulnerability using the 2016 dataset, this notebook investigates changes across time to provide additional insight into neighbourhood socioeconomic trends and the stability of the identified relationships.

In [2]:
# Import libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

pd.set_option("display.max_columns", None)

### Section-2 Load datasets: 2011 & 2016

Load the 2011 and 2016 machine learning-ready datasets and perform an initial validation of their structure. This step confirms that both datasets are available, have compatible dimensions, and contain the variables required for the temporal comparison conducted in this notebook.

In [3]:
# Load the machine learning-ready datasets

df_2011 = pd.read_csv(
    "../../data/processed/neighbourhood_profiles_2011_ml_ready.csv")

df_2016 = pd.read_csv(
    "../../data/processed/neighbourhood_profiles_2016_ml_ready.csv")

print("=" * 60)
print("2011 Dataset")
print("=" * 60)
print(f"Shape: {df_2011.shape}")
display(df_2011.head())

print("=" * 60)
print("2016 Dataset")
print("=" * 60)
print(f"Shape: {df_2016.shape}")
display(df_2016.head())

2011 Dataset
Shape: (140, 12)


,Neighbourhood,Population,Participation_Rate,Employment_Rate,Unemployment_Rate,Bachelors_Degree,No_Diploma,Visible_Minority,Renters,Immigrant_Population,Lone_Parent_Families,Low_Income_AfterTax_Pct
0,Agincourt North,30279.0,57.9,51.3,11.3,3530.0,5965.0,27025.0,1805.0,21490.0,1595.0,18.8
1,Agincourt South-Malvern West,21988.0,59.3,53.0,10.7,2520.0,4070.0,17780.0,1640.0,14225.0,1245.0,17.9
2,Alderwood,11904.0,66.0,61.1,7.4,1170.0,2290.0,1955.0,1025.0,4165.0,710.0,9.4
3,Annex,29177.0,71.8,66.6,7.3,6040.0,1690.0,6010.0,9780.0,8560.0,870.0,20.3
4,Banbury-Don Mills,26918.0,61.3,57.0,7.1,4370.0,2300.0,11930.0,4730.0,12765.0,1165.0,12.8


2016 Dataset
Shape: (141, 14)


,Neighbourhood,Population,Participation_Rate,Employment_Rate,Unemployment_Rate,Bachelors_Degree,No_Diploma,Immigrant_Population,Visible_Minority,Renters,Core_Housing_Need,Unaffordable_Housing,Lone_Parent_Families,Low_Income_AfterTax_Pct
0,City of Toronto,2731571,64.7,59.3,8.2,534610,561090,1266005,1385855,525835,240780,36.6,152595,20.2
1,Agincourt North,29113,55.4,50.0,9.8,4380,7460,19995,26365,1720,2220,34.7,1650,21.3
2,Agincourt South-Malvern West,23757,59.0,53.2,9.8,4210,6090,15095,20155,2250,1795,38.2,1360,22.6
3,Alderwood,12054,66.5,62.4,6.1,1660,2960,3970,2490,950,540,21.7,670,8.2
4,Annex,30526,70.6,65.8,6.7,9135,4270,8270,7270,9870,2420,40.4,895,18.3


### Remove the City of Aggragate row

Remove the City of Toronto aggregate record from the 2016 dataset prior to the temporal comparison. This ensures that both datasets represent only neighbourhood-level observations and prevents the city-wide summary from influencing comparisons between 2011 and 2016.

In [4]:
# Remove the City of Toronto aggregate row

df_2016 = (
    df_2016[
        df_2016["Neighbourhood"] != "City of Toronto"    ]
    .reset_index(drop=True))

print("2011 Dataset Shape:", df_2011.shape)
print("2016 Dataset Shape:", df_2016.shape)

2011 Dataset Shape: (140, 12)
2016 Dataset Shape: (140, 14)


## Interpretation

The City of Toronto aggregate record was removed from the 2016 dataset to ensure consistency with the 2011 dataset, which contains only neighbourhood-level observations. Following this step, both datasets contain 140 neighbourhoods, providing a consistent basis for temporal comparison and subsequent merging.

### Section-3. Dataset compatibility assessment

Assess the compatibility of the 2011 and 2016 machine learning-ready datasets by comparing their variables and neighbourhood coverage. This validation ensures that both datasets share a common analytical foundation and can be reliably merged for the temporal comparison performed in this notebook.

In [5]:
# Compare dataset variables

variables_2011 = set(df_2011.columns)
variables_2016 = set(df_2016.columns)

common_variables = sorted(
    variables_2011.intersection(variables_2016))

only_2011 = sorted(
    variables_2011 - variables_2016)

only_2016 = sorted(
    variables_2016 - variables_2011)

print("=" * 60)
print("Common Variables")
print("=" * 60)
print(common_variables)

print("\n")

print("=" * 60)
print("Variables Only in 2011")
print("=" * 60)
print(only_2011)

print("\n")

print("=" * 60)
print("Variables Only in 2016")
print("=" * 60)
print(only_2016)

Common Variables
['Bachelors_Degree', 'Employment_Rate', 'Immigrant_Population', 'Lone_Parent_Families', 'Low_Income_AfterTax_Pct', 'Neighbourhood', 'No_Diploma', 'Participation_Rate', 'Population', 'Renters', 'Unemployment_Rate', 'Visible_Minority']


Variables Only in 2011
[]


Variables Only in 2016
['Core_Housing_Need', 'Unaffordable_Housing']


### Interpretation

The comparison confirmed that the reconstructed 2011 machine learning-ready dataset is highly compatible with the 2016 dataset. Both datasets contain identical neighbourhood identifiers and a consistent set of socioeconomic indicators, including all independent variables and the dependent variable used throughout this study.

The only variables present in the 2016 dataset but unavailable in the 2011 Toronto Neighbourhood Profiles are **Core_Housing_Need** and **Unaffordable_Housing**. These indicators were introduced in the 2016 dataset and therefore cannot be included in the temporal comparison. Consequently, subsequent comparative analyses will be performed using the common variables available in both datasets to ensure methodological consistency.

### Section-4. Selection of Common Variables

Identify the socioeconomic indicators that are common to both the 2011 and 2016 machine learning-ready datasets. Restricting the analysis to shared variables ensures that all temporal comparisons are performed using directly comparable measures and avoids bias introduced by variables available in only one census year.

In [7]:
# Table of common variables

common_variables_df = pd.DataFrame({
    "Common Variables": common_variables})

print("=" * 60)
print("Common Variables Used for Temporal Analysis")
print("=" * 60)

display(common_variables_df)

Common Variables Used for Temporal Analysis


,Common Variables
0,Bachelors_Degree
1,Employment_Rate
2,Immigrant_Population
3,Lone_Parent_Families
4,Low_Income_AfterTax_Pct
5,Neighbourhood
6,No_Diploma
7,Participation_Rate
8,Population
9,Renters


### Interpretation

A total of **12 common variables** were identified between the 2011 and 2016 machine learning-ready datasets, including the neighbourhood identifier, **10 independent variables**, and the **dependent variable** (`Low_Income_AfterTax_Pct`).

These shared variables form the analytical foundation for the temporal comparison conducted throughout this notebook. Restricting subsequent analyses to the common variables ensures that observed differences between 2011 and 2016 reflect changes in neighbourhood socioeconomic conditions rather than differences in data availability or indicator definitions.

### Section-5. Descriptive Comparison of Socioeconomic Indicators

Compare the descriptive statistics of the common socioeconomic indicators between the 2011 and 2016 Toronto Neighbourhood Profiles. This comparison provides an initial overview of how neighbourhood characteristics changed over time before conducting formal statistical analyses.

In [8]:
# Descriptive statistics comparison

analysis_variables = [
    column for column in common_variables
    if column != "Neighbourhood"]

summary_2011 = (
    df_2011[analysis_variables]
    .describe()
    .round(2))

summary_2016 = (
    df_2016[analysis_variables]
    .describe()
    .round(2))

print("=" * 60)
print("2011 Descriptive Statistics")
print("=" * 60)
display(summary_2011)

print("=" * 60)
print("2016 Descriptive Statistics")
print("=" * 60)
display(summary_2016)

2011 Descriptive Statistics


,Bachelors_Degree,Employment_Rate,Immigrant_Population,Lone_Parent_Families,Low_Income_AfterTax_Pct,No_Diploma,Participation_Rate,Population,Renters,Unemployment_Rate,Visible_Minority
count,140.00,140.00,140.00,140.00,140.00,140.00,140.00,140.00,140.00,140.00,140.00
mean,2585.61,58.50,8936.75,1050.11,18.77,2718.96,64.45,18679.00,3395.61,9.37,9024.21
std,1822.62,6.35,6043.56,630.04,7.45,1857.22,5.69,9100.83,2393.91,2.62,7976.66
min,225.00,46.50,1855.00,215.00,3.50,335.00,53.80,6488.00,185.00,5.00,1000.00
25%,1486.25,53.95,4745.00,608.75,14.25,1442.50,59.88,11853.75,1808.75,7.40,3258.75
50%,2220.00,57.65,7362.50,847.50,17.80,2307.50,63.85,16364.50,2972.50,8.95,6272.50
75%,3086.25,62.92,11127.50,1360.00,23.00,3617.50,68.20,22407.50,4162.50,11.00,11518.75
max,13825.00,81.00,30500.00,3325.00,45.90,9295.00,85.20,53350.00,13635.00,17.10,39345.00


2016 Descriptive Statistics


,Bachelors_Degree,Employment_Rate,Immigrant_Population,Lone_Parent_Families,Low_Income_AfterTax_Pct,No_Diploma,Participation_Rate,Population,Renters,Unemployment_Rate,Visible_Minority
count,140.00,140.00,140.00,140.00,140.00,140.00,140.00,140.00,140.00,140.00,140.00
mean,3819.00,59.27,9042.89,1090.39,19.51,4007.61,64.55,19511.22,3755.68,8.30,9899.04
std,3046.74,6.52,6141.42,660.31,7.89,2281.24,6.11,10033.59,3008.98,1.90,8355.39
min,445.00,47.30,1795.00,220.00,4.50,1190.00,52.20,6577.00,280.00,4.50,1100.00
25%,2053.75,54.68,4621.25,640.00,14.10,2372.50,59.88,12019.50,1946.25,6.90,3633.75
50%,3257.50,58.55,7502.50,870.00,18.55,3330.00,63.60,16749.50,3232.50,8.20,7122.50
75%,4385.00,64.30,11327.50,1365.00,23.95,5197.50,69.12,23854.50,4338.75,9.62,12505.00
max,27280.00,82.70,31040.00,3460.00,45.50,13170.00,86.60,65913.00,23930.00,14.60,40510.00


In [9]:
comparison_summary = pd.DataFrame({

    "Variable": analysis_variables,

    "2011 Mean": [
        df_2011[col].mean()
        for col in analysis_variables    ],

    "2016 Mean": [
        df_2016[col].mean()
        for col in analysis_variables    ]})

comparison_summary["Absolute Change"] = (
    comparison_summary["2016 Mean"] -
    comparison_summary["2011 Mean"])

comparison_summary["Percent Change (%)"] = (
    comparison_summary["Absolute Change"] /
    comparison_summary["2011 Mean"] * 100)

display(
    comparison_summary.round(2))

,Variable,2011 Mean,2016 Mean,Absolute Change,Percent Change (%)
0,Bachelors_Degree,2585.61,3819.00,1233.39,47.70
1,Employment_Rate,58.50,59.27,0.77,1.32
2,Immigrant_Population,8936.75,9042.89,106.14,1.19
3,Lone_Parent_Families,1050.11,1090.39,40.29,3.84
4,Low_Income_AfterTax_Pct,18.77,19.51,0.75,3.98
5,No_Diploma,2718.96,4007.61,1288.64,47.39
6,Participation_Rate,64.45,64.55,0.10,0.15
7,Population,18679.00,19511.22,832.22,4.46
8,Renters,3395.61,3755.68,360.07,10.60
9,Unemployment_Rate,9.37,8.30,-1.07,-11.39


### Interpretation

The descriptive comparison identified several notable changes in neighbourhood socioeconomic conditions between 2011 and 2016. The largest increases were observed for **Bachelor's_Degree** (+47.70%) and **No_Diploma** (+47.39%), reflecting changes in the reported educational attainment counts across Toronto neighbourhoods. Increases were also observed in **Visible_Minority** (+9.69%), **Renters** (+10.60%), **Population** (+4.46%), **Lone_Parent_Families** (+3.84%), **Immigrant_Population** (+1.19%), and **Low_Income_AfterTax_Pct** (+3.98%).

Labour market indicators remained relatively stable over the five-year period. **Employment_Rate** increased slightly (+1.32%), **Participation_Rate** changed very little (+0.15%), while **Unemployment_Rate** decreased by **11.39%**, indicating an overall improvement in neighbourhood labour market conditions.

These descriptive findings provide an initial quantitative overview of socioeconomic changes between 2011 and 2016 and establish the foundation for the formal statistical analyses presented in the following sections.

### Section-6 Statistical Comparison of Socioeconomic Indicators

Evaluate whether the observed differences in neighbourhood socioeconomic indicators between 2011 and 2016 are statistically significant. Formal hypothesis testing is performed to distinguish meaningful temporal changes from variation that may have occurred by chance.

In [11]:
from scipy.stats import ttest_rel
import numpy as np

results = []

for variable in analysis_variables:

    mean_2011 = df_2011[variable].mean()
    mean_2016 = df_2016[variable].mean()

    mean_change = mean_2016 - mean_2011

    if mean_change > 0:
        direction = "Increased"
    elif mean_change < 0:
        direction = "Decreased"
    else:
        direction = "No Change"

    t_stat, p_value = ttest_rel(
        df_2011[variable],
        df_2016[variable]    )

    results.append({

        "Variable": variable,
        "2011 Mean": mean_2011,
        "2016 Mean": mean_2016,
        "Mean Change": mean_change,
        "Direction": direction,
        "t Statistic": t_stat,
        "P-value": p_value,
        "Significant": p_value < 0.05    })

t_test_results = (
    pd.DataFrame(results)
    .round(3))

display(t_test_results)

,Variable,2011 Mean,2016 Mean,Mean Change,Direction,t Statistic,P-value,Significant
0,Bachelors_Degree,2585.607,3819.000,1233.393,Increased,-10.618,0.000,True
1,Employment_Rate,58.499,59.272,0.773,Increased,-4.046,0.000,True
2,Immigrant_Population,8936.750,9042.893,106.143,Increased,-1.205,0.230,False
3,Lone_Parent_Families,1050.107,1090.393,40.286,Increased,-5.827,0.000,True
4,Low_Income_AfterTax_Pct,18.767,19.514,0.746,Increased,-3.891,0.000,True
5,No_Diploma,2718.964,4007.607,1288.643,Increased,-11.194,0.000,True
6,Participation_Rate,64.452,64.548,0.096,Increased,-0.534,0.594,False
7,Population,18679.000,19511.221,832.221,Increased,-4.067,0.000,True
8,Renters,3395.607,3755.679,360.071,Increased,-4.097,0.000,True
9,Unemployment_Rate,9.371,8.304,-1.067,Decreased,8.256,0.000,True


### Interpretation

The paired t-test evaluated whether neighbourhood socioeconomic indicators changed significantly between 2011 and 2016. At the 5% significance level (*p* < 0.05), **9 of the 11 research variables** exhibited statistically significant changes over the study period.

Most socioeconomic indicators showed statistically significant **increases**, including **Bachelor's_Degree** (+1233.39), **No_Diploma** (+1288.64), **Population** (+832.22), **Visible_Minority** (+874.82), **Renters** (+360.07), **Employment_Rate** (+0.77), **Lone_Parent_Families** (+40.29), and **Low_Income_AfterTax_Pct** (+0.75). In contrast, **Unemployment_Rate** was the only variable that demonstrated a statistically significant **decrease** (−1.07), indicating an improvement in neighbourhood labour market conditions.

Two variables, **Immigrant_Population** (*p* = 0.230) and **Participation_Rate** (*p* = 0.594), did not exhibit statistically significant changes between 2011 and 2016.

Overall, the paired t-test provides strong statistical evidence that most neighbourhood socioeconomic indicators changed significantly over the five-year period, supporting the conclusion that Toronto neighbourhoods experienced measurable socioeconomic changes between 2011 and 2016.

### Section-7. Neighbourhood-Level Change Analysis

Quantify neighbourhood-level changes in socioeconomic indicators between 2011 and 2016. This analysis identifies the magnitude and direction of change for each neighbourhood, highlighting areas that experienced the greatest socioeconomic shifts during the study period.

In [13]:
# Neighbourhood-level change in low-income prevalence

change_df = pd.DataFrame({

    "Neighbourhood": df_2011["Neighbourhood"],

    "Low_Income_2011":
        df_2011["Low_Income_AfterTax_Pct"],

    "Low_Income_2016":
        df_2016["Low_Income_AfterTax_Pct"]})

# Calculate change
change_df["Change (2016 - 2011)"] = (
    change_df["Low_Income_2016"] -
    change_df["Low_Income_2011"])

# Largest increases
largest_increases = (
    change_df
    .sort_values(
        by="Change (2016 - 2011)",
        ascending=False    )
    .head(10))

# Largest decreases
largest_decreases = (
    change_df
    .sort_values(
        by="Change (2016 - 2011)",
        ascending=True    )
    .head(10))

print("=" * 60)
print("Largest Increases in Low-Income Prevalence")
print("=" * 60)

display(largest_increases.round(2))

print("=" * 60)
print("Largest Decreases in Low-Income Prevalence")
print("=" * 60)

display(largest_decreases.round(2))

Largest Increases in Low-Income Prevalence


,Neighbourhood,Low_Income_2011,Low_Income_2016,Change (2016 - 2011)
130,Willowdale West,19.3,27.1,7.8
7,Bayview Village,16.4,23.7,7.3
107,Rustic,25.3,31.7,6.4
118,Thorncliffe Park,39.1,45.5,6.4
6,Bay Street Corridor,31.6,37.7,6.1
126,Weston,23.6,29.6,6.0
129,Willowdale East,24.9,30.3,5.4
13,Black Creek,28.1,33.0,4.9
1,Agincourt South-Malvern West,17.9,22.6,4.7
52,Hillcrest Village,17.2,21.9,4.7


Largest Decreases in Low-Income Prevalence


,Neighbourhood,Low_Income_2011,Low_Income_2016,Change (2016 - 2011)
53,Humber Heights-Westmount,18.7,14.6,-4.1
36,Edenbridge-Humber Valley,14.4,10.5,-3.9
100,Regent Park,45.9,42.3,-3.6
89,North St. James Town,39.7,36.5,-3.2
27,Corso Italia-Davenport,16.4,13.4,-3.0
10,Beechborough-Greenbrook,30.9,28.0,-2.9
9,Bedford Park-Nortown,12.2,9.4,-2.8
41,Etobicoke West Mall,20.6,18.3,-2.3
106,Runnymede-Bloor West Village,9.2,6.9,-2.3
61,Kennedy Park,28.5,26.3,-2.2


### Interpretation

The neighbourhood-level analysis identified substantial variation in changes in low-income prevalence across Toronto neighbourhoods between 2011 and 2016. The largest increases were observed in **Willowdale West** (+7.8 percentage points), **Bayview Village** (+7.3), **Rustic** (+6.4), **Thorncliffe Park** (+6.4), and **Bay Street Corridor** (+6.1), indicating notable increases in neighbourhood socioeconomic vulnerability during the study period.

Conversely, several neighbourhoods experienced meaningful decreases in low-income prevalence. The largest reductions were recorded in **Humber Heights–Westmount** (−4.1 percentage points), **Edenbridge–Humber Valley** (−3.9), **Regent Park** (−3.6), **North St. James Town** (−3.2), and **Corso Italia–Davenport** (−3.0), suggesting improvements in socioeconomic conditions within these communities.

Overall, these findings demonstrate that changes in neighbourhood low-income prevalence were not uniform across Toronto. While the city-wide analyses indicated a modest overall increase in low-income prevalence, the neighbourhood-level comparison reveals considerable spatial variation, with some neighbourhoods experiencing substantial increases in socioeconomic vulnerability while others showed measurable improvements. These localized findings provide valuable context for understanding the uneven distribution of socioeconomic change across Toronto between 2011 and 2016.

### Section-8. Research Question 3 Findings

Synthesize the results of the descriptive, statistical, and neighbourhood-level analyses to answer Research Question 3 by identifying which socioeconomic indicators changed significantly between 2011 and 2016, and quantifying the magnitude and direction of those changes.

The temporal comparison demonstrated that Toronto neighbourhoods experienced measurable socioeconomic changes between 2011 and 2016.

The descriptive comparison identified increases in several neighbourhood socioeconomic indicators, while the paired t-test confirmed that **9 of the 11 research variables** changed significantly (*p* < 0.05). Statistically significant increases were observed for **Population**, **Bachelor's_Degree**, **No_Diploma**, **Visible_Minority**, **Renters**, **Employment_Rate**, **Lone_Parent_Families**, and **Low_Income_AfterTax_Pct**, whereas **Unemployment_Rate** showed a statistically significant decrease. **Participation_Rate** and **Immigrant_Population** did not exhibit statistically significant changes.

Neighbourhood-level analysis further demonstrated that these changes were not evenly distributed across Toronto. Some neighbourhoods experienced substantial increases in low-income prevalence, while others showed measurable reductions, indicating considerable spatial variation in socioeconomic change across the city.

### Answer to Research Question 3

The analyses demonstrate that most neighbourhood-level socioeconomic indicators changed significantly between 2011 and 2016. Of the **11 research variables** examined, **9 exhibited statistically significant temporal changes**, while **2 remained statistically stable**. The magnitude and direction of change varied across neighbourhoods, indicating that socioeconomic conditions evolved unevenly throughout Toronto during the study period. These findings provide quantitative evidence that neighbourhood socioeconomic change was both statistically significant and spatially heterogeneous.



### Section 9. Conclusion

This notebook conducted a temporal validation and comparative analysis of Toronto neighbourhood socioeconomic conditions using the 2011 and 2016 machine learning-ready datasets. Dataset compatibility was first confirmed by identifying a common set of **11 research variables** (10 independent variables and 1 dependent variable), providing a consistent foundation for temporal comparison.

Descriptive analyses quantified changes in neighbourhood socioeconomic indicators between 2011 and 2016, while paired t-tests demonstrated that **9 of the 11 research variables** changed significantly over the study period (*p* < 0.05). The comparison further showed that employment rates increased and unemployment rates decreased, whereas participation rate and immigrant population remained statistically stable.

Neighbourhood-level analysis revealed that socioeconomic change was not uniform across Toronto. Some neighbourhoods experienced substantial increases in low-income prevalence, while others recorded measurable decreases, demonstrating considerable spatial variation in socioeconomic vulnerability.

Overall, the analyses provide quantitative evidence that neighbourhood socioeconomic conditions evolved significantly between 2011 and 2016. These findings complement the predictive modelling results presented in earlier notebooks and provide additional insight into the magnitude, direction, and spatial distribution of socioeconomic change across Toronto neighbourhoods.